### PR TITLE
jsonize smoke effect from smoke field

### PR DESCRIPTION
**Pull request by KorGgenT**
#### Summary

```SUMMARY: Features "JSONize effects caused by fields"```
#### Purpose of change

More JSONization is better. This particular PR allows fields to have their effects be defined in json.
#### Describe the solution

adds a new struct, `field_effect_data` which contains several data points that can be used to create a new effect. the effect is applied in map::creature_in_field(), as the creature class has all of the necessary data for applying effects. It might be considered duplicate code in some respects, so if it is preferred to be split and doubled into player_in_field and monster_in_field are, it should be relatively easy to do so. In order to unhardcode a couple more fields, there will need to be additional C++.
#### Describe alternatives you've considered

hardcode every new field i want to create for magiclysm.
#### Additional context

Tested with a smokebomb

### CODE REVIEW

File: src/field.cpp [hunk 1]
```diff
@@ -262,3 +264,19 @@ int field::total_move_cost() const
     }
     return current_cost;
 }
+
+effect field_entry::field_effect() const
+{
+    const field_effect_data &field_effect = type->intensity_levels[intensity - 1].field_effect;
```
Comment (Qrox) [inline src/field.cpp]:
```suggestion
    const field_effect_data &field_effect = type->get_intensity_level( intensity - 1 ).field_effect;
```

File: src/field.cpp [hunk 2]
```diff
@@ -262,3 +264,19 @@ int field::total_move_cost() const
...
+    if( fx_id.is_empty() || fx_id.is_null() ) {
+        return effect();
+    }
+    return effect( &field_effect.id.obj(), rng( field_effect.min_duration, field_effect.max_duration ),
+                   field_effect.bp, false, field_effect.intensity, calendar::turn );
+}
+
+bool field_entry::inside_immune() const
+{
+    return type->intensity_levels[intensity].field_effect.inside_immune;
```
Comment (Qrox) [inline src/field.cpp]:
```suggestion
    return type->get_intensity_level( intensity - 1 ).field_effect.inside_immune;
```

File: src/field.h [hunk 3]
```diff
@@ -9,6 +9,8 @@
 #include "color.h"
 #include "field_type.h"
 
+class effect;
```
Comment (Qrox) [inline src/field.h]: I think this isn't needed here since `field_effect()` is already returning by value?

Comment (KorGgenT) [reply src/field.h]: this is needed, what's not needed is including the whole file.

**Comment by ZhilkinSerg** [comment]
I am still getting crashes:

image

**Comment by AMurkin** [comment]
`const effect field_fx = cur_field_entry.field_effect();`

field_effect consructed here

https://github.com/CleverRaven/Cataclysm-DDA/blob/636370985c2e7e936ab36a410b1f6e182a0e03af/src/field.cpp#L273

have eff_type = nullptr.

https://github.com/CleverRaven/Cataclysm-DDA/blob/636370985c2e7e936ab36a410b1f6e182a0e03af/src/effect.h#L145

I have this on the #34397.

**Comment by FeepingCreature** [comment]
Just ran into the same issue.

**Pull request merged**